// client/src/pages/payment/PaymentForm.jsx
import React, { useState, useEffect } from "react";
import { loadStripe } from "@stripe/stripe-js";
import {
  Elements,
  PaymentElement,
  useStripe,
  useElements,
} from "@stripe/react-stripe-js";
import { api } from "../../utils/api.js";
import "./PaymentForm.css";

const stripePromise = loadStripe(import.meta.env.VITE_API_PUBLISH_KEY);

const PaymentFormInner = ({ amount, onSuccess }) => {
  const stripe = useStripe();
  const elements = useElements();

  const [loading, setLoading] = useState(false);
  const [cardError, setCardError] = useState("");

  const handleSubmit = async (e) => {
    e.preventDefault();
    setCardError("");

    if (!stripe || !elements) return;

    setLoading(true);

    try {
      const result = await stripe.confirmPayment({
        elements,
        confirmParams: {
          return_url: `${window.location.origin}/myOrders`,
        },
        redirect: "if_required",
      });

      if (result.error) {
        setCardError(result.error.message);
      } else if (result.paymentIntent?.status === "succeeded") {
        try {
          await Promise.resolve(onSuccess?.());
        } catch (finalizeErr) {
          console.error(finalizeErr);
          setCardError(
            finalizeErr?.response?.data?.message ||
              "Payment succeeded, but we could not complete your order. Please notify staff.",
          );
        }
      }
    } catch (err) {
      console.error(err);
      setCardError("An error occurred while processing your payment.");
    } finally {
      setLoading(false);
    }
  };

  return (
    <div className={`payment_container${loading ? " payment_container--busy" : ""}`}>
      {loading ? (
        <div
          className="payment-loading-shade"
          role="presentation"
          aria-hidden
        />
      ) : null}
      <form onSubmit={handleSubmit} className="payment-forms">
        <h3>Secure Checkout</h3>
        
        <PaymentElement />

        {cardError && <p className="error">{cardError}</p>}

        <button
          type="submit"
          disabled={!stripe || loading}
          className="pay-btn"
        >
          {loading ? "Processing..." : `Pay SAR ${(amount / 100).toFixed(2)}`}
        </button>
      </form>
    </div>
  );
};

const StripePayment = ({ amount, onSuccess }) => {
  const [clientSecret, setClientSecret] = useState("");
  const [error, setError] = useState("");
  const [initLoading, setInitLoading] = useState(true);

  useEffect(() => {
    let active = true;
    const fetchIntent = async () => {
      try {
        setInitLoading(true);
        setError("");
        const { data } = await api.post("/api/payment/create-payment-intent", {
          amount,
        });
        if (active) {
          if (data.clientSecret) {
            setClientSecret(data.clientSecret);
          } else {
            setError("Could not initiate payment.");
          }
        }
      } catch (err) {
        console.error(err);
        if (active) {
          setError("Failed to initialize payment gateway.");
        }
      } finally {
        if (active) {
          setInitLoading(false);
        }
      }
    };
    fetchIntent();
    return () => {
      active = false;
    };
  }, [amount]);

  if (initLoading) {
    return (
      <div className="payment-init-loading">
        <div className="payment-spinner" />
        <p>Loading secure payment gateway...</p>
      </div>
    );
  }

  if (error) {
    return (
      <div className="payment-init-error">
        <p className="error">{error}</p>
      </div>
    );
  }

  const options = {
    clientSecret,
    appearance: {
      theme: "night",
      variables: {
        colorPrimary: "#f0b429",
        colorBackground: "#111827",
        colorText: "#f3f4f6",
        colorDanger: "#ef4444",
        fontFamily: "Outfit, Inter, system-ui, sans-serif",
        borderRadius: "10px",
      },
    },
  };

  return (
    <Elements stripe={stripePromise} options={options}>
      <PaymentFormInner amount={amount} onSuccess={onSuccess} />
    </Elements>
  );
};

export default StripePayment;
